extensions/amp-date-countdown/1.0/date-countdown.js

```diff
@@ -47,95 +50,104 @@ const TimeUnit = {
 };
 
 // Default prop values
-const DEFAULT_OFFSET_SECONDS = 0;
 const DEFAULT_LOCALE = 'en';
 const DEFAULT_WHEN_ENDED = 'stop';
 const DEFAULT_BIGGEST_UNIT = 'DAYS';
 
+/**
+ * @param {!JsonObject} data
+ * @return {string}
+ */
+const DEFAULT_RENDER = (data) =>
+  /** @type {string} */ (`${data['days']} ${data['dd']}, ` +
+    `${data['hours']} ${data['hh']}, ` +
+    `${data['minutes']} ${data['mm']}, ` +
+    `${data['seconds']} ${data['ss']}, `);
+
 /**
  * @param {!DateCountdownPropsDef} props
  * @return {PreactDef.Renderable}
  */
 export function DateCountdown({
-  endDate,
-  timeleftMs,
-  timestampMs,
-  timestampSeconds,
-  offsetSeconds = DEFAULT_OFFSET_SECONDS,
+  dateTime,
   whenEnded = DEFAULT_WHEN_ENDED,
   locale = DEFAULT_LOCALE,
   biggestUnit = DEFAULT_BIGGEST_UNIT,
-  render,
-  children,
+  render = DEFAULT_RENDER,
   ...rest
 }) {
   useResourcesNotify();
   const {playable} = useAmpContext();
-  const epoch = useMemo(
-    () =>
-      getEpoch(endDate, timeleftMs, timestampMs, timestampSeconds) +
-      offsetSeconds * MILLISECONDS_IN_SECOND,
-    [endDate, timeleftMs, timestampMs, timestampSeconds, offsetSeconds]
+
+  const epoch = getDate(dateTime);
+
+  // How much time is left in our countdown
+  const setTimeleft = useState(epoch - Date.now())[1];
+
+  // One time calculation of time labels in specified locale
+  const localeStrings = useMemo(
+    () => getLocaleWord(/** @type {string} */ (locale)),
+    [locale]
   );
-  const [timeleft, setTimeleft] = useState(epoch - Date.now());
-  const localeStrings = useMemo(() => getLocaleWord(locale), [locale]);
+
+  // Reference to DOM element to get access to correct window
   const rootRef = useRef(null);
 
+  // A reference to our 'data', 'data' is used to populate the template
+  // A reference is used so we can use the same object instance between
+  // successive re-renders
+  const dataRef = useRef(
```

Comment:
I made a diagram in the design doc that shows the `render` order.
https://docs.google.com/document/d/1INUhhNEyVXArtNCuAG-y9evhV_3QoNwRf0GmrGTfEDE/edit#bookmark=id.m0gs38nvpn2h

We need to re-use the object instance here because in box 6, if we use a new instance of the object, it actually triggers box 3's, `useLayoutEffect`.  (this would be the `data` variable in `useRenderer`).  This would cause an infinite loop.

Reply:
nit: `rendererDataRef`

Reply:
Thanks Dima, Jake.  I've updated this to `useMemo` in the latest revision and renamed to `data` to be consistent with `date-display`